src/cart/src/cart/crt_group.c

```diff
@@ -1638,45 +1539,33 @@ crt_hdlr_uri_lookup(crt_rpc_t *rpc_req)
 		D_GOTO(out, rc);
 	}
 	ul_out->ul_uri = cached_uri;
+	ul_out->ul_tag = ul_in->ul_tag;
+
 	if (ul_out->ul_uri != NULL)
 		D_GOTO(out, rc);
 
+	/* If this server does not know rank:0 then return error */
+	if (ul_in->ul_tag == 0)
+		D_GOTO(out, rc = -DER_OOG);
+
 	/**
-	 * step 2, make sure URI of tag 0 is in cache in preparation to forward
-	 * the lookup RPC
+	 * step 2, if rank:tag is not found, lookup rank:tag=0
 	 */
-	if (ul_in->ul_tag != 0) {
-		/* tag != 0, check if tag 0 is in cache */
-		rc = crt_grp_lc_lookup(grp_priv_primary, crt_ctx->cc_idx,
+	ul_out->ul_tag = 0;
+	rc = crt_grp_lc_lookup(grp_priv_primary, crt_ctx->cc_idx,
 				       g_rank, 0, &cached_uri, NULL);
-		if (rc != 0) {
-			D_ERROR("crt_grp_lc_lookup(grp %s, rank %d, tag %d) "
-				"failed, rc: %d.\n",
-				grp_priv_primary->gp_pub.cg_grpid,
-				g_rank, 0, rc);
-			D_GOTO(out, rc);
-		}
+	if (rc != 0) {
+		D_ERROR("crt_grp_lc_lookup(grp %s, rank %d, tag %d) "
+			"failed, rc: %d.\n",
+			grp_priv_primary->gp_pub.cg_grpid,
+			g_rank, 0, rc);
+		D_GOTO(out, rc = -DER_OOG);
```

Comment:
Always DER_OOG or something else?  What if rc was -DER_NO_MEM

Reply:
Lookup actually no longer performs any allocations and doesn't/can't fail with rc. I ll change lookup to be void in next patch along with other changes.

Reply:
I like that kind of change ;)  Error-path handing is notoriously hard to test and there's a lot of error-paths that exist and we have code for like this one that can never actually be invoked anyway.